### ISSUE TITLE
Bed deformation models need to use the load averaged over the duration of a time step

### ISSUE THREAD

**Issue by ckhroulev** [pism/pism#525]
**Description**

Bed deformation models need to use the load (ice thickness) averaged over the duration of a time step of a bed deformation component.

A computationally expensive workaround: use synchronous coupling between mass continuity and bed deformation, i.e. bed deformation models take time steps as short as the rest of PISM.

**Additional context**

Whenever PISM takes short (in this context: shorter than a year) time steps *and* the climatic forcing resolves the annual cycle the modeled evolution of the ice thickness may contain a "high frequency" signal due to the SMB. For example: near a "steady state" the ice thickness would reach maximum in late winter and minimum in late summer, then increase towards the maximum over next winter and so on. This signal is especially noticeable when the system is near a "steady state" since it is not masked by a long term trend.

The current code (which uses _instantaneous snapshots_ of ice thickness at the end of the current time step) suffers from aliasing of this high frequency signal. This can lead to all kinds of strange behavior, e.g.:

- assuming constant year long time steps starting on January 1 and ending on December 31 (northern hemisphere): PISM's bed deformation code will use ice thickness near the **maximum** of the annual cycle _for each time step_
- assuming constant year long time steps starting on October 1 and ending on September 30 (northern hemisphere): PISM's bed deformation code will use ice thickness  near the **minimum** of the annual cycle _for each time step_.

The difference between corresponding loads may be small, but its influence can add up over time.

In most simulations time step lengths are chosen adaptively according to (CFL-based and SIA-diffusivity-based) stability criteria. This implies that any parameter change affecting ice speed, reporting frequency, etc may have an unpredictable and hard to understand influence on bed deformation.

**Comment by bueler** [comment]
Why not have an accumulating-load variable, updated at each short time step, to approximate the time-integral of the load, which is then divided by the long time step to get the average load (and zeroed-out) at each bed deformation computation?

**Comment by ckhroulev** [comment]
That's exactly what I have in mind.

**Comment by ckhroulev** [comment]
I managed to reproduce this by setting up 3 short runs

1. Start with ice thickness N=10m and SMB of +10, -10, -10, +10 m/year for months 1-3, 4-6, 7-9, 10-12 (using the 360 day calendar, i.e. all months are 30 days long),
2. Start with H=12.5m and SMB of -10, -10, +10, +10 m/year for months 1-3, 4-6, 7-9, 10-12,
3. Start with H=10m and constant SMB of 0 m/year.

All three cases have the average thickness over the year of 10m. If the bed deformation model is initialized so that it is in equilibrium with H=10m there should be no change of bed elevation as these runs go on.

The current code (`dev` branch and released versions) samples ice thickness *at the end of the year*, so it would use H=10m as the load on the bed in cases 1 and 3 and H=12.5m in case 2, resulting in bed elevation changing in time.

Averaging the load over the length of the bed deformation time step would resolve this.

Here's the script I used:

``` bash
#!/bin/bash

set -e
set -u
set -x

rm -rf delta_P_1.nc delta_P_2.nc in.nc

/usr/bin/env python3 <<EOF
import PISM
import PISM.testing as pt
import numpy as np

ctx = PISM.Context().ctx

ice_density = ctx.config().get_number("constants.ice.density")

Lx = 500e3
M = 5
grid = PISM.Grid.Shallow(ctx, Lx, Lx, 0, 0, M, M, PISM.CELL_CENTER, PISM.NOT_PERIODIC)

def forcing(values, filename):
    days_per_year = 360
    times = np.array([2, 6, 10, 14]) / 16 * days_per_year
    time_bounds = np.array([0, 4, 8, 12, 16]) / 16 * days_per_year

    pt.create_forcing(grid,
                      filename,
                      "delta_P",
                      "kg / (m^2 360 day)",
                      values,
                      time_units="days since 1-1-1",
                      times=times,
                      time_bounds=time_bounds)


def input_data(thickness, filename):
    bed = PISM.Scalar(grid, "bed")
    # well above sea level to avoid "flooding" a part of it and losing SMB (it is not applied
    # to ice free ocean)
    bed.set(1000.0)
    bed.metadata(0).standard_name("bedrock_altitude").units("m")

    thk = PISM.Scalar(grid, "thk")
    thk.set(thickness)
    thk.metadata(0).standard_name("land_ice_thickness").units("m")

    try:
        f = PISM.util.prepare_output(filename, calendar="360_day")
        thk.write(f)
        bed.write(f)
    finally:
        f.close()

forcing(np.array([10.0, -10, -10.0, 10.0]) * ice_density, "dP_PMMP.nc")
forcing(np.array([-10, -10, 10, 10]) * ice_density, "dP_MMPP.nc")
forcing(np.array([0, 0, 0, 0]) * ice_density, "dP_0.nc")

input_data(10, "H-10.nc")
input_data(12.5, "H-12.5.nc")
EOF


common_options="
-atmosphere uniform,delta_P
-atmosphere.delta_P.periodic
-atmosphere.uniform.precipitation 0
-bed_def lc
-bed_deformation.lc.update_interval 360days
-bed_deformation.update_interval 360days
-bootstrap
-calendar 360_day
-energy none
-max_dt 90days
-stress_balance none
-surface simple
-ys 1-1-1
"

extra_vars=thk,dHdt,climatic_mass_balance,mass_fluxes,dbdt,topg,viscous_bed_displacement,elastic_bed_displacement

extra="
-extra_times 90days
-extra_vars ${extra_vars}
"

suffix=${suffix:?Please set 'suffix'}

# 1. Bootstrap from the file with H=10m and run for N years with SMB=[10, -10, -10, 10]
#
# 2.1. Bootstrap from the file with H=10m. Run for 0s to get viscous and elastic bed
#      displacements corresponding to equilibrium with the ice thickness H=10m. Save to
#      H-10-full.nc
#
#      This step is needed so that in 2.2 below the bed deformation model is in
#      equilibrium with H=10m *at the beginning of the run*. Otherwise it would be in
#      equilibrium with H=12.5m and the load of 10m would result some uplift.
#
# 2.2. Bootstrap from H-12.5.nc and regrid elastic and viscous displacement from H-10-full.nc
#      and run for N years with SMB=[-10, 10, 10, -10]
#
# 3. Bootstrap from the file with H=10m. Run for N years with SMB=0.
#
# In the first two cases the SMB over 1 year is zero and ice thickness oscillates between
# 7.5 m and 12.5 m with the average of 10m. The average load is the same, but the load at
# the end of the year is different.

N=10
run_length=$(( N * 360 ))

# 1. Bootstrap from the file with H=10m and run for N years with SMB=[10, -10, -10, 10]
#
pismr \
   ${common_options} \
   ${extra} -extra_file ex_PMMP${suffix}.nc \
   -atmosphere.delta_P.file dP_PMMP.nc \
   -i H-10.nc \
   -o o_PMMP${suffix}.nc \
   -time.run_length ${run_length}days \
   ""

# 2.1. Bootstrap from the file with H=10m. Run for 0s to get viscous and elastic bed
#      displacements corresponding to equilibrium with the ice thickness H=10m. Save to
#      H-10-full.nc
#
pismr \
   ${common_options} \
   -atmosphere.delta_P.file dP_PMMP.nc \
   -i H-10.nc \
   -o H-10-full.nc \
   -time.run_length 0 \
   ""

# 2.2. Bootstrap from H-12.5.nc, then regrid elastic and viscous displacement from
#      H-10-full.nc and run for N years with SMB=[-10, -10, 10, 10]
#
pismr \
   ${common_options} \
   ${extra} -extra_file ex_MMPP${suffix}.nc \
   -atmosphere.delta_P.file dP_MMPP.nc \
   -i H-12.5.nc \
   -o o_MMPP${suffix}.nc \
   -regrid_file H-10-full.nc \
   -regrid_vars viscous_bed_displacement,elastic_bed_displacement \
   -time.run_length ${run_length}days \
   ""

# 3. Bootstrap from the file with H=10m. Run for N years with SMB=0.
#
pismr \
   ${common_options} \
   ${extra} -extra_file ex_0${suffix}.nc \
   -atmosphere.delta_P.file dP_0.nc \
   -i H-10.nc \
   -o o_0${suffix}.nc \
   -time.run_length ${run_length}days \
   ""

```

**Comment by ckhroulev** [comment]
Fixed in `dev` by merging #553.
